File: Protocol/_theoreme_sim/src/propagation/chargeSpreadRound.js
/* Purpose: Update node charges per round based on learned neighbor estimates instead of static initialization. */

import { THEOREM_ROOT_ID } from "../core/constants.js";

/**
 * @param {{
 *  nodes:Map<number,{id:number,qTotal:number,isGateway:boolean}>,
 *  estimates:Map<number,Map<number,number>>,
 *  config:{rootSourceCharge:number,chargeDropPerHop:number,chargeSpreadFactor:number}
 * }} state
 * @returns {{updates:number}}
 */
export function applyChargeSpreadRound(state) {
  const next = new Map();

  for (const node of state.nodes.values()) {
    if (node.id === THEOREM_ROOT_ID || node.isGateway) {
      next.set(node.id, Math.max(node.qTotal, state.config.rootSourceCharge));
      continue;
    }

    const estimateMap = state.estimates.get(node.id) ?? new Map();
    let bestEstimate = 0;

    for (const estimate of estimateMap.values()) {
      bestEstimate = Math.max(bestEstimate, estimate);
    }

    const target = Math.max(0, bestEstimate - state.config.chargeDropPerHop);
    const factor = Math.min(1, Math.max(0.01, state.config.chargeSpreadFactor));
    const blended = node.qTotal + (target - node.qTotal) * factor;

    // Keep q_total monotonic to preserve theorem-oriented parent potential behavior.
    const nextCharge = Math.max(node.qTotal, blended);
    next.set(node.id, nextCharge);
  }

  let updates = 0;
  for (const node of state.nodes.values()) {
    const value = next.get(node.id);
    if (value > node.qTotal) {
      updates += 1;
    }
    node.qTotal = value;
  }

  return { updates };
}

File: Protocol/_theoreme_sim_py/src/propagation/charge_spread_round.py
"""Purpose: Per-round node charge update based on best learned neighbor estimate."""

from __future__ import annotations

from src.core.constants import THEOREM_ROOT_ID
from src.utils.math_utils import clamp


def apply_charge_spread_round(state: dict) -> dict:
    next_values: dict[int, float] = {}

    for node in state["nodes"].values():
        if node.id == THEOREM_ROOT_ID or node.is_gateway:
            next_values[node.id] = max(
                float(node.q_total),
                float(state["config"]["rootSourceCharge"]),
            )
            continue

        estimate_map = state["estimates"].get(node.id, {})
        best_estimate = 0.0
        for estimate in estimate_map.values():
            best_estimate = max(best_estimate, float(estimate))

        target = max(0.0, best_estimate - float(state["config"]["chargeDropPerHop"]))
        factor = clamp(float(state["config"]["chargeSpreadFactor"]), 0.01, 1)
        blended = float(node.q_total) + (target - float(node.q_total)) * factor
        next_values[node.id] = max(float(node.q_total), blended)

    updates = 0
    for node in state["nodes"].values():
        next_charge = next_values[node.id]
        if next_charge > float(node.q_total):
            updates += 1
        node.q_total = next_charge

    return {"updates": updates}
